cdap-app-fabric/src/main/java/io/cdap/cdap/app/runtime/DelayedProgramController.java

```diff
@@ -0,0 +1,201 @@
+    this.runId = RunIds.fromString(programRunId.getRun());
+    this.delegateFuture = new CompletableFuture<>();
+  }
+
+  /**
+   * Sets the delegate {@link ProgramController} to the one provided.
+   */
+  void setProgramController(ProgramController controller) {
+    delegateFuture.complete(controller);
+  }
+
+  /**
+   * Force this program controller into ERROR state if the delegating program controller hasn't been set.
+   */
+  void failed(Throwable t) {
+    delegateFuture.complete(new FailedProgramController(programRunId, t));
+  }
+
+  /**
+   * Returns the {@link ProgramController} being delegated to. This method will block until the delegates is available.
+   */
+  @Override
+  public ProgramController getDelegate() {
+    try {
+      return Uninterruptibles.getUninterruptibly(delegateFuture);
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public ProgramRunId getProgramRunId() {
+    return programRunId;
+  }
+
+  @Override
+  public RunId getRunId() {
+    return runId;
+  }
+
+  @Override
+  public ListenableFuture<ProgramController> suspend() {
+    return callDelegate(ProgramController::suspend);
+  }
+
+  @Override
+  public ListenableFuture<ProgramController> resume() {
+    return callDelegate(ProgramController::resume);
+  }
+
+  @Override
+  public ListenableFuture<ProgramController> stop() {
+    return callDelegate(ProgramController::stop);
+  }
+
+  @Override
+  public State getState() {
+    try {
+      return Uninterruptibles.getUninterruptibly(delegateFuture).getState();
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public Throwable getFailureCause() {
+    try {
+      return Uninterruptibles.getUninterruptibly(delegateFuture).getFailureCause();
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public Cancellable addListener(Listener listener, Executor executor) {
+    CompletableFuture<Cancellable> cancellableFuture = new CompletableFuture<>();
+    delegateFuture.whenComplete((programController, throwable) -> {
+      if (throwable == null) {
+        cancellableFuture.complete(programController.addListener(listener, executor));
+      } else {
+        cancellableFuture.completeExceptionally(throwable);
+      }
+    });
+    return () -> {
+      try {
+        Uninterruptibles.getUninterruptibly(cancellableFuture).cancel();
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    };
+  }
+
+  @Override
+  public ListenableFuture<ProgramController> command(String name, Object value) {
+    return callDelegate(controller -> controller.command(name, value));
+  }
+
+  /**
+   * Makes a method call on the delegating {@link ProgramController} when it is available.
+   */
+  private ListenableFuture<ProgramController> callDelegate(
+    Function<ProgramController, ListenableFuture<ProgramController>> callFunc
+  ) {
```

Comment:
nit: this can be moved to the line above

Reply:
I intentionally make a new line to have a clear separation between the argument and the method body.